File: cpp/dijkstra.cpp
#include "dijkstra.h"
#include <iostream>
#include <fstream>
#include <queue>
#include <limits>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

std::unordered_map<std::string, int> dijkstra(const Graph& graph, const std::string& start) {
    std::unordered_map<std::string, int> distances;
    for (const auto& node : graph.adjacency_list) {
        distances[node.first] = std::numeric_limits<int>::max();
    }
    distances[start] = 0;

    auto compare = [](const std::pair<std::string, int>& a, const std::pair<std::string, int>& b) {
        return a.second > b.second;
    };
    std::priority_queue<std::pair<std::string, int>, 
                       std::vector<std::pair<std::string, int>>, 
                       decltype(compare)> pq(compare);
    pq.push({start, 0});

    while (!pq.empty()) {
        auto [current, current_distance] = pq.top();
        pq.pop();

        if (current_distance > distances[current]) {
            continue;
        }

        if (graph.adjacency_list.find(current) != graph.adjacency_list.end()) {
            for (const auto& [neighbor, weight] : graph.adjacency_list.at(current)) {
                int distance = current_distance + weight;
                if (distance < distances[neighbor]) {
                    distances[neighbor] = distance;
                    pq.push({neighbor, distance});
                }
            }
        }
    }

    return distances;
}

Graph load_graph_from_json(const std::string& filename) {
    Graph graph;
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Error: Could not open file " << filename << std::endl;
        return graph;
    }

    json j;
    file >> j;

    for (const auto& edge : j) {
        std::string from = edge["from"];
        std::string to = edge["to"];
        int weight = edge["weight"];
        graph.adjacency_list[from][to] = weight;
        // Ensure all nodes are in the adjacency list, even if they have no outgoing edges
        graph.adjacency_list[to];
    }

    return graph;
}

int main() {
    Graph graph = load_graph_from_json("../data/sample_graph.json");
    auto distances = dijkstra(graph, "A");
    
    std::cout << "Shortest distances from A:\n";
    for (const auto& [node, distance] : distances) {
        std::cout << node << ": " << distance << std::endl;
    }
    
    return 0;
}

File: go/dijkstra.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math"
	"os"
)

type Edge struct {
	From   string `json:"from"`
	To     string `json:"to"`
	Weight int    `json:"weight"`
}

type Graph struct {
	adjacencyList map[string]map[string]int
}

func loadGraphFromJSON(filename string) (*Graph, error) {
	file, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	var edges []Edge
	err = json.Unmarshal(file, &edges)
	if err != nil {
		return nil, err
	}

	graph := &Graph{
		adjacencyList: make(map[string]map[string]int),
	}

	for _, edge := range edges {
		if graph.adjacencyList[edge.From] == nil {
			graph.adjacencyList[edge.From] = make(map[string]int)
		}
		graph.adjacencyList[edge.From][edge.To] = edge.Weight
		// Ensure all nodes are in the adjacency list
		if graph.adjacencyList[edge.To] == nil {
			graph.adjacencyList[edge.To] = make(map[string]int)
		}
	}

	return graph, nil
}

func dijkstra(graph *Graph, start string) map[string]int {
	distances := make(map[string]int)
	for node := range graph.adjacencyList {
		distances[node] = math.MaxInt32
	}
	distances[start] = 0

	// Simple implementation without priority queue for simplicity
	visited := make(map[string]bool)

	for len(visited) < len(graph.adjacencyList) {
		// Find the unvisited node with the smallest distance
		minNode := ""
		minDist := math.MaxInt32
		for node, dist := range distances {
			if !visited[node] && dist < minDist {
				minDist = dist
				minNode = node
			}
		}

		if minNode == "" {
			break
		}

		visited[minNode] = true

		for neighbor, weight := range graph.adjacencyList[minNode] {
			if !visited[neighbor] {
				newDist := distances[minNode] + weight
				if newDist < distances[neighbor] {
					distances[neighbor] = newDist
				}
			}
		}
	}

	return distances
}

func main() {
	graph, err := loadGraphFromJSON("../data/sample_graph.json")
	if err != nil {
		fmt.Printf("Error loading graph: %v\n", err)
		os.Exit(1)
	}

	distances := dijkstra(graph, "A")
	fmt.Println("Shortest distances from A:")
	for node, distance := range distances {
		fmt.Printf("%s: %d\n", node, distance)
	}
}
